src/libp2p/collection.rs

```diff
@@ -397,6 +397,21 @@ where
         self.connections.len()
     }
 
+    /// Returns the state of the given connection.
+    ///
+    /// # Panic
+    ///
+    /// Panics if the identifier is invalid or corresponds to a connection that has already
+    /// entirely shut down.
+    ///
+    pub fn connection_state(&self, connection_id: ConnectionId) -> ConnectionState {
+        let connection = self.connections.get(&connection_id).unwrap();
```

Comment:
in my experience it's better to return an error or Option rather than panic.

Reply:
I know many people say that, but in my opinion it's not.

If you take for example the `queue_notification` function, which can fail because the queue is full, and imagine that instead the function returns an `Err` either if the queue is full or if the `ConnectionId` is wrong.

If you call `queue_notification` and get an error, what do you do? The queue being full is a legitimate problem that you can handle by buffering the notification to send it later for example. But the `ConnectionId` being invalid can't be handled properly. The only reason why you would pass an invalid `ConnectionId` is a bug in the code, and in that situation the right thing to do is to panic.

Now, in principle I agree that it's the code that calls `queue_notification` that should match the error and handle some kind of `ConnectionIdInvalid` error and call `panic!()`, but in practice this is extremely tedious. If you specifically filter out the `ConnectionIdInvalid` error, then you have to redefine another error type that contains all the variants except for `ConnectionIdInvalid`. In practice what happens is that programmers don't make the effort or simply don't think about it and propagate the `ConnectionIdInvalid` error to upper layers of the code, which is non-sensical.
Basically, it is much easier for everyone if it's `queue_notification` that panics if the `ConnectionId` is invalid.

My implicit rule is that if the only reason why an error can happen is a bug in the code, then there's no reason to properly return an error and panicking is completely appropriate.

Reply:
Sorry if my comment was too generic. I agree that in cases of bugs we should panic. But in cases such as queue being full we should instead provide back-pressure or drop incoming messages and/or figure out if we can speed up processing (whatever is more appropriate in a particular case). That being said, I think we're on the same page generally.